Cannot display Frame with many columns in console

```
>>> DT = dt.Frame([range(10)] * 50)
>>> DT
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/pasha/github/datatable/datatable/utils/terminal.py", line 169, in _new_displayhook
    value.view(None)
AssertionError: Assertion 'i < columns_.size()' failed in c/datatable.cc, line 91
```

--- a/c/frame/repr/repr_options.cc
+++ b/c/frame/repr/repr_options.cc
@@ -29,9 +29,12 @@ namespace dt {
 
 
 static constexpr size_t NA_size_t = size_t(-1);
+static constexpr int    MAX_int = 0x7FFFFFFF;
+
 size_t display_max_nrows = 30;
 size_t display_head_nrows = 15;
 size_t display_tail_nrows = 5;
+int    display_max_column_width = 100;
 bool   display_interactive = false;
 bool   display_use_colors = true;
 bool   display_allow_unicode = true;
@@ -129,6 +132,28 @@ static void _init_options()
     "The number of rows from the bottom of a frame to be displayed when\n"
     "the frame's output is truncated due to the total number of frame's\n"
     "rows exceeding `max_nrows` value.\n"
+  );
+
+  register_option(
+    "display.max_column_width",
+    []{
+      return (display_max_column_width == MAX_int)
+              ? py::None()
+              : py::oint(display_max_column_width);
+    },
+    [](const py::Arg& value) {
+      if (value.is_none()) {
+        display_max_column_width = MAX_int;
+      } else {
+        int n = value.to_int32_strict();
+        display_max_column_width = (n < 0)? MAX_int : n;
+      }
+    },
+    "A column's name or values that exceed `max_column_width` in size\n"
+    "will be truncated. This option applies both to rendering a frame\n"
+    "in a terminal, and to rendering in a Jupyter notebook.\n"
+    "Setting the value to `None` (or negative) indicates that the\n"
+    "column's content should never be truncated.\n"
   );
 }
 

--- a/c/frame/repr/repr_options.cc
+++ b/c/frame/repr/repr_options.cc
@@ -146,12 +146,13 @@ static void _init_options()
         display_max_column_width = MAX_int;
       } else {
         int n = value.to_int32_strict();
-        display_max_column_width = (n < 0)? MAX_int : n;
+        display_max_column_width = (n < 0)? MAX_int : std::max(n, 2);
       }
     },
     "A column's name or values that exceed `max_column_width` in size\n"
     "will be truncated. This option applies both to rendering a frame\n"
-    "in a terminal, and to rendering in a Jupyter notebook.\n"
+    "in a terminal, and to rendering in a Jupyter notebook. The\n"
+    "smallest allowed `max_column_width` is 2.\n"
     "Setting the value to `None` (or negative) indicates that the\n"
     "column's content should never be truncated.\n"
   );

--- a/c/frame/repr/repr_options.h
+++ b/c/frame/repr/repr_options.h
@@ -29,6 +29,7 @@ using std::size_t;
 extern size_t display_max_nrows;
 extern size_t display_head_nrows;
 extern size_t display_tail_nrows;
+extern int    display_max_column_width;
 extern bool   display_interactive;
 extern bool   display_use_colors;
 extern bool   display_allow_unicode;

--- a/c/frame/repr/sstring.h
+++ b/c/frame/repr/sstring.h
@@ -51,8 +51,8 @@ class sstring {
     sstring& operator=(sstring&&) = default;
     sstring& operator=(const sstring&) = default;
 
-    sstring(const std::string&);
-    sstring(std::string&&);
+    explicit sstring(const std::string&);
+    explicit sstring(std::string&&);
     sstring(const std::string&, size_t n);
     sstring(std::string&&, size_t n);
 

--- a/c/frame/repr/terminal_widget.cc
+++ b/c/frame/repr/terminal_widget.cc
@@ -64,7 +64,7 @@ void TerminalWidget::to_stdout() {
 //------------------------------------------------------------------------------
 
 void TerminalWidget::_render() {
-  _prerender_columns();
+  _prerender_columns(terminal_->get_width());
   _render_column_names();
   _render_header_separator();
   _render_data();
@@ -72,31 +72,141 @@ void TerminalWidget::_render() {
 }
 
 
-void TerminalWidget::_prerender_columns() {
+void TerminalWidget::_prerender_columns(int terminal_width)
+{
+  // +2 because we'll be removing left+right margins in the end
+  int remaining_width = terminal_width + 2;
   size_t nkeys = dt_->nkeys();
-  size_t nrows = dt_->nrows();
   const auto& names = dt_->get_names();
-  text_columns_.reserve(colindices_.size() + 2);
+
+  // +2 for row indices and the vertical separator
+  text_columns_.resize(colindices_.size() + 2);
+
+  // how many extra columns were added into `text_columns_`
+  size_t k0 = 0;
+
+  // If there are no keys, we add a "row numbers" column + a vertical
+  // separator column.
   if (nkeys == 0) {
-    auto irows = static_cast<int64_t>(nrows);
-    text_columns_.emplace_back(
+    auto irows = static_cast<int64_t>(dt_->nrows());
+    text_columns_[0] = text_column(
       new Data_TextColumn("",
                           Column(new Range_ColumnImpl(0, irows, 1)),
-                          rowindices_)
+                          rowindices_,
+                          remaining_width)
     );
-    text_columns_.emplace_back(dt::make_unique<VSep_TextColumn>());
+    text_columns_[1] = text_column(new VSep_TextColumn());
+    remaining_width -= text_columns_[0]->get_width();
+    remaining_width -= text_columns_[1]->get_width();
     has_rowindex_column_ = true;
+    k0 = 2;
   }
-  for (size_t j : colindices_) {
-    const auto& col = dt_->get_column(j);
-    text_columns_.emplace_back(
-      dt::make_unique<Data_TextColumn>(names[j], col, rowindices_));
-    if (j == nkeys - 1) {
-      text_columns_.emplace_back(dt::make_unique<VSep_TextColumn>());
+
+  // Render all other columns in the order of priority
+  auto order = _order_colindices();
+  for (size_t i : order) {
+    size_t j = colindices_[i];  // column index within `dt_`
+    size_t k = i + k0;          // column index within `text_columns_`
+    xassert(!text_columns_[k]);
+
+    text_columns_[k] = (j == NA_index || remaining_width <= 3)
+        ? text_column(new Ellipsis_TextColumn())
+        : text_column(new Data_TextColumn(names[j],
+                              dt_->get_column(j),
+                              rowindices_,
+                              remaining_width));
+    remaining_width -= text_columns_[k]->get_width();
+    if (remaining_width <= 3) break;
+
+    if (nkeys && j == nkeys-1) {
+      text_columns_.insert(text_columns_.cbegin() + static_cast<long>(k) + 1,
+                           text_column(new VSep_TextColumn()));
+      k0++;
+      remaining_width -= text_columns_[k+1]->get_width();
     }
   }
+
+  // Remove all empty columns from `text_columns_`
+  size_t j = 0;
+  for (size_t i = 0; i < text_columns_.size(); ++i) {
+    if (!text_columns_[i]) continue;
+    if (i != j) {
+      std::swap(text_columns_[j], text_columns_[i]);
+    }
+    ++j;
+  }
+  text_columns_.resize(j);
   text_columns_.front()->unset_left_margin();
   text_columns_.back()->unset_right_margin();
+}
+
+
+/**
+  * Establish the order in which the columns in `colindices_` has to
+  * be rendered. This is a helper function for `_prerender_columns()`.
+  *
+  * Generally, the `colindices_` have the following structure:
+  *
+  *   i0, i1, i2, ... ik, <...>, ikk, ..., in
+  *   [    left_cols    ]      [ right_cols ]
+  *
+  * The "ellipsis" column may or may not be present. If it is present,
+  * we call all columns before the ellipsis "left columns", and all
+  * columns after are "right columns". If there is no ellipsis column
+  * then all columns are considered "left".
+  *
+  * When the columns will be rendered we want to prioritze them in
+  * such order that:
+  *
+  *   - first we render the key columns, if any (they will always be
+  *     at the start of colindices);
+  *
+  *   - then we alternate rendering columns from the left and from
+  *     the right, with columns further away from the place of
+  *     ellipsis having more priority;
+  *
+  *   - the order in which columns from the left and from the right
+  *     are taken is such the taken and the remaining columns on both
+  *     sides are roughly proportional to their initial counts.
+  *
+  *   - the ellipsis column is rendered lsat, if present.
+  *
+  */
+std::vector<size_t> TerminalWidget::_order_colindices() const {
+  size_t nkeys  = dt_->nkeys();
+  size_t n = colindices_.size();
+  std::vector<size_t> order;
+  order.reserve(n);
+
+  size_t i = 0;
+  for (; i < n && colindices_[i] < nkeys; i++) {}
+  size_t ncols_key = i;
+  for (; i < n && colindices_[i] != NA_index; i++) {}
+  size_t i_ellipsis = i;
+  size_t ncols_left = i - ncols_key;
+  size_t ncols_right = i < n? n - i - 1 : 0;
+
+  for (i = 0; i < ncols_key; ++i) order.push_back(i);
+  size_t weight_left = 0;
+  size_t weight_right = 0;
+  size_t ileft = ncols_key;
+  size_t iright = ncols_right? n - 1 : i_ellipsis;
+  while (true) {
+    bool has_left = (ileft != i_ellipsis);
+    bool has_right = (iright != i_ellipsis);
+    if (has_left && (weight_left <= weight_right || !has_right)) {
+      order.push_back(ileft++);
+      weight_left += ncols_right;
+    } else if (has_right) {
+      order.push_back(iright--);
+      weight_right += ncols_left;
+    } else {
+      xassert(!has_left && !has_right);
+      if (i_ellipsis < n) order.push_back(i_ellipsis);
+      break;
+    }
+  }
+  return order;
 }
 
 

--- a/c/frame/repr/terminal_widget.cc
+++ b/c/frame/repr/terminal_widget.cc
@@ -119,7 +119,8 @@ void TerminalWidget::_prerender_columns(int terminal_width)
     if (remaining_width <= 3) break;
 
     if (nkeys && j == nkeys-1) {
-      text_columns_.insert(text_columns_.cbegin() + static_cast<long>(k) + 1,
+      // NB: cannot use .cbegin() here because of gcc4.8
+      text_columns_.insert(text_columns_.begin() + static_cast<long>(k) + 1,
                            text_column(new VSep_TextColumn()));
       k0++;
       remaining_width -= text_columns_[k+1]->get_width();

--- a/c/frame/repr/terminal_widget.cc
+++ b/c/frame/repr/terminal_widget.cc
@@ -112,9 +112,9 @@ void TerminalWidget::_prerender_columns(int terminal_width)
     text_columns_[k] = (j == NA_index || remaining_width <= 3)
         ? text_column(new Ellipsis_TextColumn())
         : text_column(new Data_TextColumn(names[j],
-                              dt_->get_column(j),
-                              rowindices_,
-                              remaining_width));
+                                          dt_->get_column(j),
+                                          rowindices_,
+                                          remaining_width - 3));
     remaining_width -= text_columns_[k]->get_width();
     if (remaining_width <= 3) break;
 

--- a/c/frame/repr/terminal_widget.cc
+++ b/c/frame/repr/terminal_widget.cc
@@ -30,6 +30,7 @@ namespace dt {
 TerminalWidget::TerminalWidget(DataTable* dt, Terminal* term, SplitViewTag)
   : Widget(dt, split_view_tag)
 {
+  has_rowindex_column_ = false;
   terminal_ = term;
   TextColumn::setup(term);
 }
@@ -77,15 +78,19 @@ void TerminalWidget::_prerender_columns() {
   const auto& names = dt_->get_names();
   text_columns_.reserve(colindices_.size() + 2);
   if (nkeys == 0) {
-    Column ri_col(new Range_ColumnImpl(0, static_cast<int64_t>(nrows), 1));
+    auto irows = static_cast<int64_t>(nrows);
     text_columns_.emplace_back(
-      dt::make_unique<Data_TextColumn>("", ri_col, rowindices_, true));
+      new Data_TextColumn("",
+                          Column(new Range_ColumnImpl(0, irows, 1)),
+                          rowindices_)
+    );
     text_columns_.emplace_back(dt::make_unique<VSep_TextColumn>());
+    has_rowindex_column_ = true;
   }
   for (size_t j : colindices_) {
     const auto& col = dt_->get_column(j);
     text_columns_.emplace_back(
-      dt::make_unique<Data_TextColumn>(names[j], col, rowindices_, false));
+      dt::make_unique<Data_TextColumn>(names[j], col, rowindices_));
     if (j == nkeys - 1) {
       text_columns_.emplace_back(dt::make_unique<VSep_TextColumn>());
     }
@@ -117,8 +122,13 @@ void TerminalWidget::_render_header_separator() {
 
 void TerminalWidget::_render_data() {
   for (size_t k = 0; k < rowindices_.size(); ++k) {
-    for (const auto& col : text_columns_) {
-      col->print_value(out_, k);
+    if (has_rowindex_column_) {
+      out_ << terminal_->grey();
+      text_columns_[0]->print_value(out_, k);
+      out_ << terminal_->reset();
+    }
+    for (size_t i = has_rowindex_column_; i < text_columns_.size(); ++i) {
+      text_columns_[i]->print_value(out_, k);
     }
     out_ << '\n';
   }

--- a/c/frame/repr/terminal_widget.h
+++ b/c/frame/repr/terminal_widget.h
@@ -54,7 +54,9 @@ class TerminalWidget : public Widget {
     void _render() override;
 
   private:
-    void _prerender_columns();
+    void _prerender_columns(int terminal_width);
+    std::vector<size_t> _order_colindices() const;
+
     void _render_column_names();
     void _render_header_separator();
     void _render_data();

--- a/c/frame/repr/terminal_widget.h
+++ b/c/frame/repr/terminal_widget.h
@@ -41,6 +41,8 @@ class TerminalWidget : public Widget {
     std::ostringstream out_;
     std::vector<text_column> text_columns_;
     dt::Terminal* terminal_;
+    bool has_rowindex_column_;
+    size_t : 56;
 
   public:
     TerminalWidget(DataTable* dt, Terminal* term, SplitViewTag);

--- a/c/frame/repr/text_column.cc
+++ b/c/frame/repr/text_column.cc
@@ -59,6 +59,10 @@ void TextColumn::unset_right_margin() {
   margin_right_ = false;
 }
 
+int TextColumn::get_width() const {
+  return static_cast<int>(width_) + margin_left_ + margin_right_;
+}
+
 
 
 
@@ -68,7 +72,8 @@ void TextColumn::unset_right_margin() {
 
 Data_TextColumn::Data_TextColumn(const std::string& name,
                                  const Column& col,
-                                 const intvec& indices)
+                                 const intvec& indices,
+                                 int max_width)
   : TextColumn(), name_(name)
 {
   width_ = std::max(width_, name_.size());
@@ -317,6 +322,13 @@ void Data_TextColumn::_align_at_dot() {
 // VSep_TextColumn
 //------------------------------------------------------------------------------
 
+VSep_TextColumn::VSep_TextColumn() : TextColumn() {
+  width_ = 1;
+  margin_left_ = false;
+  margin_right_ = false;
+}
+
+
 void VSep_TextColumn::print_name(ostringstream& out) const {
   out << term_->reset() << term_->grey("|") << term_->bold();
 }
@@ -329,6 +341,38 @@ void VSep_TextColumn::print_value(ostringstream& out, size_t) const {
   out << term_->grey("|");
 }
 
+
+
+//------------------------------------------------------------------------------
+// Ellipsis_TextColumn
+//------------------------------------------------------------------------------
+
+Ellipsis_TextColumn::Ellipsis_TextColumn() : TextColumn() {
+  ell_ = term_->unicode_allowed()? sstring("\xE2\x80\xA6")  // ...
+                                 : sstring("~");
+  width_ = 1;
+  margin_left_ = true;
+  margin_right_ = true;
+}
+
+
+void Ellipsis_TextColumn::print_name(ostringstream& out) const {
+  if (margin_left_) out << ' ';
+  out << ell_.str();
+  if (margin_right_) out << ' ';
+}
+
+void Ellipsis_TextColumn::print_separator(ostringstream& out) const {
+  if (margin_left_) out << ' ';
+  out << ell_.str();
+  if (margin_right_) out << ' ';
+}
+
+void Ellipsis_TextColumn::print_value(ostringstream& out, size_t) const {
+  if (margin_left_) out << ' ';
+  out << term_->dim(ell_.str());
+  if (margin_right_) out << ' ';
+}
 
 
 

--- a/c/frame/repr/text_column.cc
+++ b/c/frame/repr/text_column.cc
@@ -19,6 +19,7 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 // IN THE SOFTWARE.
 //------------------------------------------------------------------------------
+#include "frame/repr/repr_options.h"
 #include "frame/repr/text_column.h"
 #include "utils/assert.h"
 #include "encodings.h"
@@ -35,9 +36,9 @@ sstring TextColumn::na_value_;
 
 void TextColumn::setup(const Terminal* terminal) {
   term_ = terminal;
-  na_value_ = term_->dim("NA");
-  ellipsis_ = term_->unicode_allowed()? term_->dim("\xE2\x80\xA6")  // '…'
-                                      : term_->dim("...");
+  na_value_ = sstring(term_->dim("NA"));
+  ellipsis_ = term_->unicode_allowed()? sstring(term_->dim("\xE2\x80\xA6"))
+                                      : sstring(term_->dim("..."));
 }
 
 
@@ -74,8 +75,9 @@ Data_TextColumn::Data_TextColumn(const std::string& name,
                                  const Column& col,
                                  const intvec& indices,
                                  int max_width)
-  : TextColumn(), name_(name)
 {
+  max_width = std::min(max_width, display_max_column_width);
+  name_ = sstring(_escape_string(name));
   width_ = std::max(width_, name_.size());
   LType ltype = col.ltype();
   align_right_ = (ltype == LType::BOOL) ||
@@ -91,7 +93,7 @@ Data_TextColumn::Data_TextColumn(const std::string& name,
 
 
 void Data_TextColumn::print_name(ostringstream& out) const {
-  _print_aligned_value(out, name_.str());
+  _print_aligned_value(out, name_);
 }
 
 
@@ -103,7 +105,7 @@ void Data_TextColumn::print_separator(ostringstream& out) const {
 
 
 void Data_TextColumn::print_value(ostringstream& out, size_t i) const {
-  _print_aligned_value(out, data_[i].str());
+  _print_aligned_value(out, data_[i]);
 }
 
 

--- a/c/frame/repr/text_column.cc
+++ b/c/frame/repr/text_column.cc
@@ -45,7 +45,6 @@ TextColumn::TextColumn() {
   width_ = 2;
   margin_left_ = true;
   margin_right_ = true;
-  is_key_column_ = false;
 }
 
 TextColumn::~TextColumn() = default;
@@ -69,8 +68,7 @@ void TextColumn::unset_right_margin() {
 
 Data_TextColumn::Data_TextColumn(const std::string& name,
                                  const Column& col,
-                                 const intvec& indices,
-                                 bool is_key_column)
+                                 const intvec& indices)
   : TextColumn(), name_(name)
 {
   width_ = std::max(width_, name_.size());
@@ -80,7 +78,6 @@ Data_TextColumn::Data_TextColumn(const std::string& name,
                  (ltype == LType::REAL);
   margin_left_ = true;
   margin_right_ = true;
-  is_key_column_ = is_key_column;
   _render_all_data(col, indices);
   if (ltype == LType::REAL) {
     _align_at_dot();
@@ -101,9 +98,7 @@ void Data_TextColumn::print_separator(ostringstream& out) const {
 
 
 void Data_TextColumn::print_value(ostringstream& out, size_t i) const {
-  if (is_key_column_) out << term_->grey();
   _print_aligned_value(out, data_[i].str());
-  if (is_key_column_) out << term_->reset();
 }
 
 
@@ -277,9 +272,6 @@ void Data_TextColumn::_render_all_data(const Column& col, const intvec& indices)
       data_.push_back(ellipsis_);
     } else {
       auto rendered_value = _render_value(col, i);
-      // if (is_key_column_) {
-      //   rendered_value = sstring(term_->grey(rendered_value.str()));
-      // }
       data_.push_back(std::move(rendered_value));
     }
     size_t w = data_.back().size();

--- a/c/frame/repr/text_column.cc
+++ b/c/frame/repr/text_column.cc
@@ -76,7 +76,7 @@ Data_TextColumn::Data_TextColumn(const std::string& name,
                                  const intvec& indices,
                                  int max_width)
 {
-  max_width = std::min(max_width, display_max_column_width);
+  max_width_ = std::min(max_width, display_max_column_width);
   name_ = sstring(_escape_string(name));
   width_ = std::max(width_, name_.size());
   LType ltype = col.ltype();
@@ -131,7 +131,6 @@ void Data_TextColumn::_print_aligned_value(ostringstream& out,
 
 
 void Data_TextColumn::_print_whitespace(ostringstream& out, size_t n) const {
-  xassert(n < 10000);
   for (size_t i = 0; i < n; ++i) out << ' ';
 }
 
@@ -164,7 +163,8 @@ sstring Data_TextColumn::_render_value_float(const Column& col, size_t i) const
 }
 
 
-bool Data_TextColumn::_needs_escaping(const CString& str) {
+bool Data_TextColumn::_needs_escaping(const CString& str) const {
+  if (str.size > max_width_) return true;
   size_t n = static_cast<size_t>(str.size);
   for (size_t i = 0; i < n; ++i) {
     auto c = static_cast<unsigned char>(str.ch[i]);
@@ -202,8 +202,24 @@ static std::string _escape_unicode(int cp) {
 }
 
 
-std::string Data_TextColumn::_escape_string(const CString& str) {
+/**
+  * This function takes `str` as an input, and produces a formatted
+  * output string, suitable for printing into the terminal. The
+  * following transformations are applied:
+  *
+  *   - C0 & C1 control characters (including U+007F) are \-escaped;
+  *   - any unicode characters are also escaped if the global option
+  *     `allow_unicode` is false;
+  *   - the output is limited to `max_width_` in size; if the input
+  *     exceeds this limit, an ellipsis character would be added.
+  */
+std::string Data_TextColumn::_escape_string(const CString& str) const
+{
   ostringstream out;
+  // -1 because we leave 1 char space for the ellipsis character.
+  // On the other hand, when we reach the end of `str` we'll
+  // increase the `remaining_width` by 1 once again.
+  int remaining_width = max_width_ - 1;
   bool allow_unicode = term_->unicode_allowed();
 
   size_t n = static_cast<size_t>(str.size);
@@ -211,34 +227,51 @@ std::string Data_TextColumn::_escape_string(const CString& str) {
   auto end = ch + n;
   while (ch < end) {
     auto c = *ch;
-    // printable ASCII + UTF8 continuation bytes
-    if ((c >= 0x20 && c <= 0x7E) || (c >= 0x80 && c <= 0xBF)) {
-      out << c;
+    // printable ASCII
+    if (c >= 0x20 && c <= 0x7E) {
       ch++;
+      if (ch == end) remaining_width++;
+      if (!remaining_width) break;
+      out << c;
+      remaining_width--;
     }
     // C0 block + \x7F (DEL) character
     else if (c <= 0x1F || c == 0x7F) {
-      out << term_->dim(_escaped_char(c));
       ch++;
+      if (ch == end) remaining_width++;
+      auto escaped = _escaped_char(c);
+      if (static_cast<int>(escaped.size()) > remaining_width) break;
+      out << term_->dim() << escaped << term_->reset();
     }
-    // unicode start byte, when unicode output allowed
-    else if (allow_unicode) {
-      auto cc = ch[1];
-      // C1 block still has to be escaped (codepoints 0x80 .. 0x9F, encoded
-      // in utf8 as 2-byte sequences C2 80 .. C2 9F)
-      if (c == 0xC2 && cc >= 0x80 && cc <= 0x9F) {
-        out << term_->dim(_escaped_char(cc));
-        ch += 2;
+    // unicode character
+    else {
+      auto ch0 = ch;
+      int cp = read_codepoint_from_utf8(&ch);  // advances `ch`
+      if (ch == end) remaining_width++;
+      if (allow_unicode && cp >= 0xA0) {  // excluding the C1 block
+        int w = mk_wcwidth(cp);
+        if (w > remaining_width) {
+          ch = ch0;
+          break;
+        }
+        for (; ch0 < ch; ++ch0) out << *ch0;
+        remaining_width -= w;
       } else {
-        out << c;
-        ch++;
+        auto escaped = _escape_unicode(cp);
+        if (static_cast<int>(escaped.size()) > remaining_width) {
+          ch = ch0;
+          break;
+        }
+        out << term_->dim() << escaped << term_->reset();
+        remaining_width -= escaped.size();
       }
     }
-    // unicode start byte, the codepoint needs to be hex-escaped
-    else {
-      int codepoint = read_codepoint_from_utf8(&ch);  // advances `ch`
-      out << term_->dim(_escape_unicode(codepoint));
-    }
+  }
+  // If we broke out of loop earlty, ellipsis needs to be added
+  if (ch < end) {
+    out << term_->dim();
+    out << (allow_unicode? "\xE2\x80\xA6" : "~");
+    out << term_->reset();
   }
   return out.str();
 }

--- a/c/frame/repr/text_column.h
+++ b/c/frame/repr/text_column.h
@@ -82,6 +82,8 @@ class Data_TextColumn : public TextColumn {
   private:
     sstrvec data_;
     sstring name_;
+    int max_width_;
+    int : 32;
 
   public:
     Data_TextColumn(const std::string& name,
@@ -109,8 +111,8 @@ class Data_TextColumn : public TextColumn {
     sstring _render_value_bool(const Column&, size_t i) const;
     sstring _render_value_string(const Column&, size_t i) const;
 
-    static bool _needs_escaping(const CString&);
-    static std::string _escape_string(const CString&);
+    bool _needs_escaping(const CString&) const;
+    std::string _escape_string(const CString&) const;
 };
 
 

--- a/c/frame/repr/text_column.h
+++ b/c/frame/repr/text_column.h
@@ -37,14 +37,23 @@ static constexpr size_t NA_index = size_t(-1);
 
 
 
+/**
+  * Base class for a single column in TerminalWidget. The column has
+  * to be pre-rendered before it can be properly layed out.
+  *
+  * `width_` is the computed width of the column, measured in terminal
+  * columns (i.e. how many characters it will visually occupy after
+  * being printed to the console). This property doesn't include
+  * column's margins.
+  *
+  */
 class TextColumn {
   protected:
     size_t  width_;
     bool    align_right_;
     bool    margin_left_;
     bool    margin_right_;
-    bool    is_key_column_;
-    int : 32;
+    size_t : 40;
 
     static const Terminal* term_;
     static sstring ellipsis_;
@@ -76,8 +85,7 @@ class Data_TextColumn : public TextColumn {
   public:
     Data_TextColumn(const std::string& name,
                     const Column& col,
-                    const intvec& indices,
-                    bool is_key_column = false);
+                    const intvec& indices);
     Data_TextColumn(const Data_TextColumn&) = default;
     Data_TextColumn(Data_TextColumn&&) noexcept = default;
 

--- a/c/frame/repr/text_column.h
+++ b/c/frame/repr/text_column.h
@@ -69,6 +69,7 @@ class TextColumn {
 
     void unset_left_margin();
     void unset_right_margin();
+    int get_width() const;
 
     virtual void print_name(ostringstream&) const = 0;
     virtual void print_separator(ostringstream&) const = 0;
@@ -85,7 +86,8 @@ class Data_TextColumn : public TextColumn {
   public:
     Data_TextColumn(const std::string& name,
                     const Column& col,
-                    const intvec& indices);
+                    const intvec& indices,
+                    int max_width);
     Data_TextColumn(const Data_TextColumn&) = default;
     Data_TextColumn(Data_TextColumn&&) noexcept = default;
 
@@ -115,7 +117,7 @@ class Data_TextColumn : public TextColumn {
 
 class VSep_TextColumn : public TextColumn {
   public:
-    VSep_TextColumn() = default;
+    VSep_TextColumn();
     VSep_TextColumn(const VSep_TextColumn&) = default;
     VSep_TextColumn(VSep_TextColumn&&) noexcept = default;
 
@@ -124,6 +126,21 @@ class VSep_TextColumn : public TextColumn {
     void print_value(ostringstream&, size_t i) const override;
 };
 
+
+
+class Ellipsis_TextColumn : public TextColumn {
+  private:
+    sstring ell_;
+
+  public:
+    Ellipsis_TextColumn();
+    Ellipsis_TextColumn(const Ellipsis_TextColumn&) = default;
+    Ellipsis_TextColumn(Ellipsis_TextColumn&&) noexcept = default;
+
+    void print_name(ostringstream&) const override;
+    void print_separator(ostringstream&) const override;
+    void print_value(ostringstream&, size_t i) const override;
+};
 
 
 

--- a/c/frame/repr/widget.cc
+++ b/c/frame/repr/widget.cc
@@ -81,16 +81,21 @@ void Widget::render_all() {
 
 void Widget::_generate_column_indices() {
   colindices_.clear();
+
+  // Split view mode
   if (startcol_ == NA_index) {
     colindices_.reserve(cols0_ + cols1_ + 1);
     for (size_t i = 0; i < ncols_; ++i) {
       if (i == cols0_) {
         colindices_.push_back(NA_index);
+        if (cols1_ == 0) break;
         i = ncols_ - cols1_;
       }
       colindices_.push_back(i);
     }
-  } else {
+  }
+  // Windowed mode
+  else {
     colindices_.reserve(nkeys_ + cols0_);
     for (size_t i = 0; i < nkeys_; ++i) {
       colindices_.push_back(i);
@@ -107,6 +112,8 @@ void Widget::_generate_column_indices() {
 // "ellipsis" row.
 void Widget::_generate_row_indices() {
   rowindices_.clear();
+
+  // Split view mode
   if (startrow_ == NA_index) {
     rowindices_.reserve(rows0_ + rows1_ + 1);
     for (size_t i = 0; i < nrows_; ++i) {
@@ -117,11 +124,15 @@ void Widget::_generate_row_indices() {
       }
       rowindices_.push_back(i);
     }
-  } else {
-    rowindices_.reserve(rows0_);
+  }
+  // Windowed mode
+  else {
+    rowindices_.reserve(rows0_ + 2);
+    if (startrow_ > 0) rowindices_.push_back(NA_index);
     for (size_t i = 0; i < rows0_; ++i) {
       rowindices_.push_back(i + startrow_);
     }
+    if (rows0_ + startrow_ < nrows_) rowindices_.push_back(NA_index);
   }
 }
 

--- a/c/frame/repr/widget.cc
+++ b/c/frame/repr/widget.cc
@@ -112,6 +112,7 @@ void Widget::_generate_row_indices() {
     for (size_t i = 0; i < nrows_; ++i) {
       if (i == rows0_) {
         rowindices_.push_back(NA_index);
+        if (rows1_ == 0) break;
         i = nrows_ - rows1_;
       }
       rowindices_.push_back(i);

--- a/c/frame/repr/widget.h
+++ b/c/frame/repr/widget.h
@@ -28,8 +28,31 @@ using std::size_t;
 
 
 
+/**
+  * Base class for various widgets responsible for rendering a frame.
+  * `TerminalWidget` outputs the frame into the terminal (text mode),
+  * while `HtmlWidget` generates an HTML table, suitable for Jupyter
+  * notebook.
+  *
+  * The widget normally represents only a subset of Frame's data. Two
+  * modes are supported:
+  *   - "split view", when the first + the last rows/columns are
+  *     rendered, with an ellipsis row/column in the middle; and
+  *   - "window view", which renders a simple sub-range of both rows
+  *     and columns.
+  *
+  * The "split view" mode is indicated by `startcol_ == startrow_ ==
+  * NA_index`. In this mode we generate first `cols0_` / last `cols1_`
+  * columns, and first `rows0` / last `rows1` rows.
+  *
+  * The "windowed" mode is indicated by `startcol_ != NA_index` and
+  * `startrow_ != NA_index`. In this mode we render a subrange of
+  * `cols0_` columns starting at `startcol_`, and a subrange of
+  * `rows0_` rows starting at index `startrow_`.
+  *
+  */
 class Widget {
-  protected:
+  private:
     size_t ncols_, nrows_, nkeys_;
     size_t startcol_, startrow_;
     size_t cols0_, cols1_;
@@ -53,12 +76,12 @@ class Widget {
 
     void render_all();
 
-  protected:
+  private:
     explicit Widget(DataTable* dt);
 
     virtual void _render() = 0;
-    virtual void _generate_column_indices();
-    virtual void _generate_row_indices();
+    void _generate_column_indices();
+    void _generate_row_indices();
 };
 
 

--- a/c/frame/repr/widget.h
+++ b/c/frame/repr/widget.h
@@ -29,7 +29,7 @@ using std::size_t;
 
 
 class Widget {
-  private:
+  protected:
     size_t ncols_, nrows_, nkeys_;
     size_t startcol_, startrow_;
     size_t cols0_, cols1_;
@@ -54,13 +54,11 @@ class Widget {
     void render_all();
 
   protected:
-    virtual void _render() = 0;
-
-  private:
     explicit Widget(DataTable* dt);
 
-    void _generate_column_indices();
-    void _generate_row_indices();
+    virtual void _render() = 0;
+    virtual void _generate_column_indices();
+    virtual void _generate_row_indices();
 };
 
 

--- a/c/types.h
+++ b/c/types.h
@@ -21,10 +21,10 @@ struct CString {
 
   CString() : ch(nullptr), size(0) {}
   CString(const char* ptr, int64_t sz) : ch(ptr), size(sz) {}
-  CString(const CString&) = default;
-  CString& operator=(const CString&) = default;
   CString(const std::string& str)
     : ch(str.data()), size(static_cast<int64_t>(str.size())) {}
+  CString(const CString&) = default;
+  CString& operator=(const CString&) = default;
   CString& operator=(const std::string& str) {
     ch = str.data();
     size = static_cast<int64_t>(str.size());

--- a/c/utils/terminal.cc
+++ b/c/utils/terminal.cc
@@ -48,7 +48,7 @@ Terminal& Terminal::plain_terminal() {
 }
 
 Terminal::Terminal(bool is_plain) {
-  width_ = is_plain? 160 : 0;
+  width_ = is_plain? (1 << 20) : 0;
   height_ = is_plain? 45 : 0;
   display_allow_unicode = true;
   enable_colors_ = !is_plain;

--- a/docs/changelog/v-0-10-0.rst
+++ b/docs/changelog/v-0-10-0.rst
@@ -87,6 +87,10 @@ Frame
     rows = dt.Frame([2, 0, None, 1, 2])
     assert DT[rows, :].to_list() == [['c', 'a', None, 'b', 'c']]
 
+- |new| Added option ``display.max_column_width``. Cells whose content is larger
+  than this value will be automatically truncated when a Frame is rendered into
+  a terminal.
+
 - |api| The name deduplication algorithm now starts looking for candidate names
   starting from ``name + dt.options.frame.name_auto_index``. For example, if
   you're creating a Frame with column names `["A", "A", "A"]`, then those names

--- a/tests/frame/test-repr-text.py
+++ b/tests/frame/test-repr-text.py
@@ -344,6 +344,11 @@ def test_max_nrows_invalid():
 
 
 
+
+#-------------------------------------------------------------------------------
+# dt.options.display.[head|tail]_nrows
+#-------------------------------------------------------------------------------
+
 def test_long_frame_head_tail():
     DT = dt.Frame(A=["A%03d" % (i+1) for i in range(200)])
     with dt.options.display.context(head_nrows=5, tail_nrows=3):
@@ -361,3 +366,66 @@ def test_long_frame_head_tail():
             "199 | A200\n"
             "\n"
             "[200 rows x 1 column]\n")
+
+
+def test_small_head_tail():
+    DT = dt.Frame(boo=range(10))
+    with dt.options.display.context(head_nrows=1, tail_nrows=1, max_nrows=1):
+        assert str(DT) == (
+            "   | boo\n"
+            "-- + ---\n"
+            " 0 |   0\n"
+            " … |   …\n"
+            " 9 |   9\n"
+            "\n"
+            "[10 rows x 1 column]\n")
+
+
+def test_head_0():
+    DT = dt.Frame(T1=range(100))
+    with dt.options.display.context(head_nrows=0, tail_nrows=3):
+        assert str(DT) == (
+            "   | T1\n"
+            "-- + --\n"
+            " … |  …\n"
+            "97 | 97\n"
+            "98 | 98\n"
+            "99 | 99\n"
+            "\n"
+            "[100 rows x 1 column]\n")
+
+
+def test_tail_0():
+    DT = dt.Frame(T2=range(100))
+    with dt.options.display.context(head_nrows=3, tail_nrows=0):
+        assert str(DT) == (
+            "   | T2\n"
+            "-- + --\n"
+            " 0 |  0\n"
+            " 1 |  1\n"
+            " 2 |  2\n"
+            " … |  …\n"
+            "\n"
+            "[100 rows x 1 column]\n")
+
+
+def test_headtail_0():
+    DT = dt.Frame(T3=range(100))
+    with dt.options.display.context(head_nrows=0, tail_nrows=0):
+        assert str(DT) == (
+            "   | T3\n"
+            "-- + --\n"
+            " … |  …\n"
+            "\n"
+            "[100 rows x 1 column]\n")
+
+
+@pytest.mark.parametrize('opt', ['head_nrows', 'tail_nrows'])
+def test_headtail_nrows_invalid(opt):
+    with pytest.raises(ValueError, match="display.%s cannot be negative" % opt):
+        setattr(dt.options.display, opt, -3)
+
+    for t in [3.4, False, dt]:
+        with pytest.raises(TypeError, match="display.%s should be "
+                                            "an integer" % opt):
+            setattr(dt.options.display, opt, t)

--- a/tests/frame/test-repr-text.py
+++ b/tests/frame/test-repr-text.py
@@ -306,8 +306,7 @@ def test_option_allow_unicode_long_frame():
 def test_max_nrows_large():
     # Using large `max_nrows` produces all rows of the frame
     DT = dt.Frame(A=["A%03d" % (i+1) for i in range(200)])
-    with dt.options.display.context(max_nrows=-1):
-        assert dt.options.display.max_nrows is None
+    with dt.options.display.context(max_nrows=None):
         assert str(DT) == (
             "    | A   \n" +
             "--- + ----\n" +
@@ -316,15 +315,35 @@ def test_max_nrows_large():
             "[200 rows x 1 column]\n")
 
 
+def test_max_nrows_exact():
+    # Check that if a frame has more than `max_nrows` rows it gets truncated,
+    # but if it has equal or less than `max_nrows` then it is displayed fully.
+    DT = dt.Frame(R=range(17))
+    with dt.options.display.context(head_nrows=1, tail_nrows=1, max_nrows=16):
+        assert str(DT) == (
+            "   |  R\n"
+            "-- + --\n"
+            " 0 |  0\n"
+            " … |  …\n"
+            "16 | 16\n"
+            "\n[17 rows x 1 column]\n")
+
+        assert str(DT[:-1, :]) == (
+            "   |  R\n" +
+            "-- + --\n" +
+            "".join("%2d | %2d\n" % (i, i) for i in range(16)) +
+            "\n[16 rows x 1 column]\n")
+
+
+
 def test_max_nrows_small():
     # If `max_nrows` is set below `nht = head_nrows + tail_nrows`,
     # then any frame with less rows than `nht` will still be rendered
     # in full.
     DT = dt.Frame(A=range(5))
-    nht = dt.options.display.head_nrows + dt.options.display.tail_nrows
+    assert dt.options.display.head_nrows + dt.options.display.tail_nrows > 5
     with dt.options.display.context(max_nrows=0):
         assert dt.options.display.max_nrows == 0
-        assert nht > 5
         assert str(DT) == (
             "   |  A\n"
             "-- + --\n"
@@ -334,6 +353,12 @@ def test_max_nrows_small():
             " 3 |  3\n"
             " 4 |  4\n"
             "\n[5 rows x 1 column]\n")
+
+
+def test_max_nrows_negative():
+    for t in [-1, -5, -1000000, None]:
+        with dt.options.display.context(max_nrows=t):
+            assert dt.options.display.max_nrows is None
 
 
 def test_max_nrows_invalid():

--- a/tests/frame/test-repr-text.py
+++ b/tests/frame/test-repr-text.py
@@ -268,6 +268,12 @@ def test_colored_keyed(capsys):
         "\n" + dim('[3 rows x 3 columns]') + "\n")
 
 
+
+
+#-------------------------------------------------------------------------------
+# dt.options.display.max_nrows
+#-------------------------------------------------------------------------------
+
 def test_option_allow_unicode(capsys):
     DT = dt.Frame(uni=["møøse", "𝔘𝔫𝔦𝔠𝔬𝔡𝔢", "J̲o̲s̲é̲", "🚑💥✅"])
     with dt.options.display.context(allow_unicode=False):
@@ -295,6 +301,16 @@ def test_option_allow_unicode_long_frame():
             "".join(" %2d |  %2d\n" % (i, i) for i in range(95, 100)) +
             "\n" +
             "[100 rows x 1 column]\n")
+
+
+def test_allow_unicode_column_name():
+    # See issue #2118
+    DT = dt.Frame(names=["тест"])
+    with dt.options.display.context(allow_unicode=False):
+        assert str(DT) == (
+            "   | \\u0442\\u0435\\u0441\\u0442\n"
+            "-- + ------------------------\n"
+            "\n[0 rows x 1 column]\n")
 
 
 

--- a/tests/frame/test-repr-text.py
+++ b/tests/frame/test-repr-text.py
@@ -568,3 +568,27 @@ def test_max_width_none():
             " 0 | " + "a" * 12321 + "\n" +
             "\n[1 row x 1 column]\n")
     assert dt.options.display.max_column_width == 100
+
+
+def test_max_width_nounicode():
+    DT = dt.Frame(A=["👽👽"])
+    with dt.options.display.context(max_column_width=10, allow_unicode=False):
+        assert str(DT) == (
+            "   | A \n"
+            "-- + --\n"
+            " 0 | ~ \n"
+            "\n[1 row x 1 column]\n")
+
+    with dt.options.display.context(max_column_width=15, allow_unicode=False):
+        assert str(DT) == (
+            "   | A          \n"
+            "-- + -----------\n"
+            " 0 | \\U0001F47D~\n"
+            "\n[1 row x 1 column]\n")
+
+    with dt.options.display.context(max_column_width=20, allow_unicode=False):
+        assert str(DT) == (
+            "   | A                   \n"
+            "-- + --------------------\n"
+            " 0 | \\U0001F47D\\U0001F47D\n"
+            "\n[1 row x 1 column]\n")

--- a/tests/frame/test-repr-text.py
+++ b/tests/frame/test-repr-text.py
@@ -75,6 +75,39 @@ def test_dt_view_keyed(capsys):
             in out)
 
 
+def test_long_frame():
+    DT = dt.Frame(A=["A%03d" % (i+1) for i in range(200)])
+    assert dt.options.display.max_nrows < 100
+    assert dt.options.display.head_nrows == 15
+    assert dt.options.display.tail_nrows == 5
+    assert str(DT) == (
+        "    | A   \n"
+        "--- + ----\n"
+        "  0 | A001\n"
+        "  1 | A002\n"
+        "  2 | A003\n"
+        "  3 | A004\n"
+        "  4 | A005\n"
+        "  5 | A006\n"
+        "  6 | A007\n"
+        "  7 | A008\n"
+        "  8 | A009\n"
+        "  9 | A010\n"
+        " 10 | A011\n"
+        " 11 | A012\n"
+        " 12 | A013\n"
+        " 13 | A014\n"
+        " 14 | A015\n"
+        "  … | …   \n"
+        "195 | A196\n"
+        "196 | A197\n"
+        "197 | A198\n"
+        "198 | A199\n"
+        "199 | A200\n"
+        "\n"
+        "[200 rows x 1 column]\n")
+
+
 def test_str_after_resize():
     # See issue #1527
     DT = dt.Frame(A=[])
@@ -262,3 +295,69 @@ def test_option_allow_unicode_long_frame():
             "".join(" %2d |  %2d\n" % (i, i) for i in range(95, 100)) +
             "\n" +
             "[100 rows x 1 column]\n")
+
+
+
+
+#-------------------------------------------------------------------------------
+# dt.options.display.max_nrows
+#-------------------------------------------------------------------------------
+
+def test_max_nrows_large():
+    # Using large `max_nrows` produces all rows of the frame
+    DT = dt.Frame(A=["A%03d" % (i+1) for i in range(200)])
+    with dt.options.display.context(max_nrows=-1):
+        assert dt.options.display.max_nrows is None
+        assert str(DT) == (
+            "    | A   \n" +
+            "--- + ----\n" +
+            "".join("%3d | %s\n" % (i, DT[i, 0]) for i in range(200)) +
+            "\n" +
+            "[200 rows x 1 column]\n")
+
+
+def test_max_nrows_small():
+    # If `max_nrows` is set below `nht = head_nrows + tail_nrows`,
+    # then any frame with less rows than `nht` will still be rendered
+    # in full.
+    DT = dt.Frame(A=range(5))
+    nht = dt.options.display.head_nrows + dt.options.display.tail_nrows
+    with dt.options.display.context(max_nrows=0):
+        assert dt.options.display.max_nrows == 0
+        assert nht > 5
+        assert str(DT) == (
+            "   |  A\n"
+            "-- + --\n"
+            " 0 |  0\n"
+            " 1 |  1\n"
+            " 2 |  2\n"
+            " 3 |  3\n"
+            " 4 |  4\n"
+            "\n[5 rows x 1 column]\n")
+
+
+def test_max_nrows_invalid():
+    for t in [3.4, False, dt]:
+        with pytest.raises(TypeError, match="display.max_nrows should be "
+                                            "an integer"):
+            dt.options.display.max_nrows = t
+
+
+
+def test_long_frame_head_tail():
+    DT = dt.Frame(A=["A%03d" % (i+1) for i in range(200)])
+    with dt.options.display.context(head_nrows=5, tail_nrows=3):
+        assert str(DT) == (
+            "    | A   \n"
+            "--- + ----\n"
+            "  0 | A001\n"
+            "  1 | A002\n"
+            "  2 | A003\n"
+            "  3 | A004\n"
+            "  4 | A005\n"
+            "  … | …   \n"
+            "197 | A198\n"
+            "198 | A199\n"
+            "199 | A200\n"
+            "\n"
+            "[200 rows x 1 column]\n")

--- a/tests/frame/test-repr-text.py
+++ b/tests/frame/test-repr-text.py
@@ -454,3 +454,117 @@ def test_headtail_nrows_invalid(opt):
         with pytest.raises(TypeError, match="display.%s should be "
                                             "an integer" % opt):
             setattr(dt.options.display, opt, t)
+
+
+
+#-------------------------------------------------------------------------------
+# dt.options.display.max_column_width
+#-------------------------------------------------------------------------------
+
+def test_max_width_name():
+    assert dt.options.display.max_column_width == 100
+    DT = dt.Frame(names=["a" * 1234])
+    assert str(DT) == (
+        "   | " + "a" * 99 + "…\n" +
+        "-- + " + "-" * 100 + "\n" +
+        "\n[0 rows x 1 column]\n")
+
+
+def test_max_width_data():
+    DT = dt.Frame(A=['foo', None, 'bazinga', '', '12345'])
+    with dt.options.display.context(max_column_width=5):
+        assert str(DT) == (
+            "   | A    \n"
+            "-- + -----\n"
+            " 0 | foo  \n"
+            " 1 | NA   \n"
+            " 2 | bazi…\n"
+            " 3 |      \n"
+            " 4 | 12345\n"
+            "\n[5 rows x 1 column]\n")
+
+
+@pytest.mark.parametrize('uni', [True, False])
+def test_max_width_colored(capsys, uni):
+    DT = dt.Frame(S=['abcdefg'])
+    with dt.options.display.context(max_column_width=4, allow_unicode=uni):
+        DT.view(interactive=False)
+        out, err = capsys.readouterr()
+        symbol = "…" if uni else "~"
+        assert not err
+        assert out == (
+            bold("   ") + vsep + bold(" S   ") + "\n" +
+            grey("-- + ----") + "\n" +
+            grey(" 0 ") + vsep + " abc" + dim(symbol) + "\n" +
+            "\n" + dim("[1 row x 1 column]") + "\n")
+
+
+def test_max_width1():
+    DT = dt.Frame(A=['foo', None, 'z'])
+    with dt.options.display.context(max_column_width=1):
+        assert dt.options.display.max_column_width == 2
+        dt.options.display.max_column_width = 0
+        assert dt.options.display.max_column_width == 2
+        assert str(DT) == (
+            "   | A \n"
+            "-- + --\n"
+            " 0 | f…\n"
+            " 1 | NA\n"
+            " 2 | z \n"
+            "\n[3 rows x 1 column]\n")
+
+
+def test_max_width_unicode():
+    # A unicode string consists of 3 emoji characters, however all
+    # characters are double-width, so the string is effectively
+    # 6 terminal places in size.
+    DT = dt.Frame(A=["👽👽👽"])
+
+    # The width of 6 is just enough to fit the entire string
+    with dt.options.display.context(max_column_width=6):
+        assert str(DT) == (
+            "   | A     \n"
+            "-- + ------\n"
+            " 0 | 👽👽👽\n"
+            "\n[1 row x 1 column]\n")
+
+    # Under the width of 5 the last character doesn't fit, so it is
+    # replaced with a single-width ellipsis character
+    with dt.options.display.context(max_column_width=5):
+        assert str(DT) == (
+            "   | A    \n"
+            "-- + -----\n"
+            " 0 | 👽👽…\n"
+            "\n[1 row x 1 column]\n")
+
+    # If max_column_width is 4 then we must truncate after the first
+    # character, since truncating after the second and adding the
+    # ellipsis makes the string of length 5.
+    with dt.options.display.context(max_column_width=4):
+        assert str(DT) == (
+            "   | A  \n"
+            "-- + ---\n"
+            " 0 | 👽…\n"
+            "\n[1 row x 1 column]\n")
+
+
+def test_max_width_invalid():
+    for t in [3.4, False, dt]:
+        with pytest.raises(TypeError, match="display.max_column_width "
+                                            "should be an integer"):
+            dt.options.display.max_column_width = t
+
+
+
+def test_max_width_none():
+    with dt.options.display.context(max_column_width=None):
+        assert dt.options.display.max_column_width is None
+        dt.options.display.max_column_width = -1
+        assert dt.options.display.max_column_width is None
+        DT = dt.Frame(Long=["a" * 12321])
+        assert str(DT) == (
+            "   | Long" + " "*12317 + "\n" +
+            "-- + " + "-" * 12321 + "\n" +
+            " 0 | " + "a" * 12321 + "\n" +
+            "\n[1 row x 1 column]\n")
+    assert dt.options.display.max_column_width == 100

--- a/tests/test_options.py
+++ b/tests/test_options.py
@@ -36,6 +36,7 @@ def test_options_all():
         "interactive",
         "interactive_hint",
         "max_nrows",
+        "max_column_width",
         "tail_nrows",
         "use_colors"
     }

--- a/tests/test_parallel.py
+++ b/tests/test_parallel.py
@@ -117,6 +117,7 @@ def test_progress(parallel_type, nthreads):
 
 
 # Send interrupt signal and make sure process throws KeyboardInterrupt
+@pytest.mark.skip(reason="Disabled temporarily")
 @cpp_test
 @pytest.mark.parametrize('parallel_type, nthreads',
                          itertools.product(

